src/modules/content-s3/s3.c

```diff
@@ -0,0 +1,259 @@
+
+#include "s3.h"
+
+static S3Status status_g;
+static S3Protocol protocol = S3ProtocolHTTP;
+static S3UriStyle uri_style = S3UriStylePath;
+
+/* Data needed by the get object callback function
+ */
+struct get_cb_data {
+    size_t size;
+    void *data;
+};
+
+struct put_cb_data {
+    const void *data;
+    int count;
+};
+
+static S3Status response_props_cb(const S3ResponseProperties *properties, void *data)
+{
+    return S3StatusOK;
+}
+
+static void response_complete_cb(S3Status status, const S3ErrorDetails *error, void *data)
+{
+    status_g = status;
+}
+
+/* Writes a chunk of 'data' to s3, returning
+ * the size of the data written.
+ */
+static int put_object_cb(int buff_size, char *buff, void *data)
+{
+    struct put_cb_data *cb_data = data;
+
+    memcpy(buff, cb_data->data + cb_data->count, buff_size);
+    cb_data->count += buff_size;
+
+    return buff_size;
+}
+
+/* Gets the object from s3 storing it in in 'data->datap' as well
+ * as updating 'data->sizep' to hold the size of the data read.
+ */
+static S3Status get_object_cb(int buff_size, const char *buff, void *data)
+{
+    void *tmp = NULL;
+    struct get_cb_data *cb_data = data;
+
+    if (!(tmp = realloc (cb_data->data, buff_size + cb_data->size))) {
+        free(tmp);
+        return S3StatusOutOfMemory;
+    }
+    cb_data->data = tmp;
+
+    memcpy(cb_data->data + cb_data->size, buff, buff_size);
+    cb_data->size += buff_size;
+
+    return S3StatusOK;
+}
+
+int s3_init(struct s3_config *cfg, const char **errstr)
+{
+    S3Status status = S3_initialize("s3", S3_INIT_ALL, cfg->hostname);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name(status);
+
+        return -1;
+    }
+    
+    return 0;
+}
+
+void s3_cleanup()
+{
+    S3_deinitialize();
+}
+
+int s3_bucket_create(struct s3_config *cfg, const char **errstr)
+{
+    int retries = cfg->retries;
+    S3Status status = S3_validate_bucket_name(cfg->bucket, uri_style);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name(status);
+
+        return -1;
+    }
+
+    S3ResponseHandler resp_hndl = {
+        &response_props_cb,
+        &response_complete_cb
+    };
+
+    do {
+        S3_create_bucket(protocol, cfg->access_key, cfg->secret_key, 0, cfg->bucket, S3CannedAclPrivate, 0, 0, &resp_hndl, 0);
+        retries--;
+    } while (S3_status_is_retryable(status_g) && retries > 0);
+
+    if (status_g != S3StatusOK && status_g != S3StatusErrorBucketAlreadyOwnedByYou) {
+        errno = EREMOTEIO;
+        if (errstr)
+            *errstr = S3_get_status_name(status);
+
+        return -1;
+    }
+
+    return 0;
+}
+
+int s3_put(struct s3_config *cfg, const char *key, const void *data, size_t size, const char **errstr)
+{
+    int retries = cfg->retries;
+    int count = 0;
+
+    S3ResponseHandler resp_hndl = {
+        &response_props_cb,
```

Comment:
It would probably be a bit clearer, and also less prone to breakage if the libs3 API ever changes, to use [tagged initializers](https://stackoverflow.com/questions/3016107/what-is-tagged-structure-initialization-syntax)  when initializing these structures (in all the functions).